### PR TITLE
Check for dependencies in west flash/debug

### PR DESCRIPTION
**Pull request by mbolivar**
The west flash/debug support for JLink checks if the required programs are missing, and prints an error message if so.

This PR factors that out into runners.core and uses it everywhere.

I've increased test coverage for runners I know well or that seem commonly used, to try to make sure this doesn't break anything. There still isn't basic run/flash/debug test coverage for every single runner, however.

### CODE REVIEW

**Comment by mbolivar** [comment]
> There still isn't basic run/flash/debug test coverage for every single runner, however.

I plan on adding openocd testing at some point, but other than that, help is wanted in cases where I don't have any hardware (arc.py, esp32.py, hifive1.py, intel_s1000.py, nios2.py, nsim.py, xtensa.py).

**Review by carlescufi — COMMENTED** [review]
I have not reviewed the tests in detail, but the new exception class and method make sense

File: scripts/west_commands/runners/core.py [hunk 1]
```diff
@@ -154,6 +156,19 @@ def _parse_value(self, value):
...
+    '''FileNotFoundError subclass for missing program dependencies.
+
+    No significant changes from the parent FileNotFoundError; this is
+    useful for explicitly signaling that the file in question is a
+    program that some class requires to proceed.
+
+    The filename attribute contains the missing program.'''
+
+    def __init__(self, program):
+        super().__init__(errno.ENOENT, os.strerror(errno.ENOENT), program)
```
Comment (carlescufi) [inline scripts/west_commands/runners/core.py]: This time you used the Python 3 style ;)

**Review by carlescufi — APPROVED** [review]
Looks good, thanks for the extensive tests!

File: scripts/west_commands/runners/core.py [hunk 2]
```diff
@@ -412,6 +427,20 @@ def do_run(self, command, **kwargs):
...
+
+        :param program: name of the program that is required,
+                        or path to a program binary.
+
+        If ``program`` is an absolute path to an existing program
+        binary, this call succeeds. Otherwise, try to find the program
+        by name on the system PATH.
+
+        On error, raises MissingProgram.'''
+        if shutil.which(program) is None:
```
[the remark this answers is not in the document]
Comment (mbolivar) [reply scripts/west_commands/runners/core.py]: Yes, it works as intended. Examples:

```
>>> import shutil
>>> shutil.which('JLink')
'C:\\Program Files (x86)\\SEGGER\\JLink_V644g\\JLink.EXE'
>>> shutil.which('C:\\Program Files (x86)\\SEGGER\\JLink_V644g\\JLink.EXE')
'C:\\Program Files (x86)\\SEGGER\\JLink_V644g\\JLink.EXE'
>>> shutil.which('JLink.EXE')
'C:\\Program Files (x86)\\SEGGER\\JLink_V644g\\JLink.EXE'
>>> shutil.which('jlink.exe')
'C:\\Program Files (x86)\\SEGGER\\JLink_V644g\\jlink.exe'
```

**Review by nashif — APPROVED** [review]

**Comment by mbolivar** [comment]
Are we OK to merge? I have more PRs I want to send that depend on this one.

**Pull request merged**